### ISSUE TITLE
Gradle Hangs While Attempting to Build and Publish Proof-of-Concept Image

### ISSUE THREAD

**Issue by neilchaudhuri** [GoogleContainerTools/jib#1964]
**Environment**:

- *Jib version:* 0.10.1
- *Build tool:* grade 5.4.1
- *OS:* Windows 10

**Description of the issue**:
Gradle hangs when attempting to build a simple "Hello Jib" image to a local, deliberately insecure registry as a proof of concept.

**Expected behavior**:
Gradle should build the image based on Redhat's OpenJDK with Jib successfully to the local, insecure repository as a proof of concept. 

**Steps to reproduce**:

  1. Create a local insecure Docker registry
  2. Use jib-core within a Gradle build (more on why below) to build an image (from any base you want) to the local registry on Windows 10

**`jib-gradle-plugin` Configuration**: 
```groovy
task(buildImage) {
  doLast {
    Jib
      .from("my-repo.com:5000/redhat-openjdk-18/openjdk18-openshift:1.6")
      .containerize(Containerizer
          .to(RegistryImage.named("localhost:5000/hello-from-jib"))
          .setAllowInsecureRegistries(true)
      )
  }
}
```

**Log output**: 

```
11:56:48.083 [INFO] [com.google.api.client.http.HttpTransport] -------------- REQUEST  --------------
GET https://localhost:5000/v2/
Accept:
Accept-Encoding: gzip
User-Agent: jib 0.10.1 jib-core Google-HTTP-Java-Client/1.30.0 (gzip)

11:56:48.084 [INFO] [com.google.api.client.http.HttpTransport] curl -v --compressed -H 'Accept: ' -H 'Accept-Encoding: gzip' -H 'User-Agent: jib 0.10.1 jib-core Google-HTTP-Java-Client/1.30.0 (gzip)' -- 'https://localhost:5000/v2/'
11:56:48.105 [DEBUG] [org.apache.http.client.protocol.RequestAddCookies] CookieSpec selected: default
11:56:48.116 [DEBUG] [org.apache.http.client.protocol.RequestAuthCache] Auth cache not set in the context
11:56:48.117 [DEBUG] [org.apache.http.impl.conn.PoolingHttpClientConnectionManager] Connection request: [route: {s}->https://localhost:5000][total kept alive: 0; route allocated: 0 of 20; total allocated: 0 of 200]
11:56:48.132 [DEBUG] [org.apache.http.impl.conn.PoolingHttpClientConnectionManager] Connection leased: [id: 0][route: {s}->https://localhost:5000][total kept alive: 0; route allocated: 1 of 20; total allocated: 1 of 200]
11:56:48.134 [DEBUG] [org.apache.http.impl.execchain.MainClientExec] Opening connection {s}->https://localhost:5000
11:56:48.146 [DEBUG] [org.apache.http.impl.conn.DefaultHttpClientConnectionOperator] Connecting to localhost/127.0.0.1:5000
11:56:48.147 [DEBUG] [org.apache.http.conn.ssl.SSLConnectionSocketFactory] Connecting socket to localhost/127.0.0.1:5000 with timeout 0

11:56:48.211 [DEBUG] [org.apache.http.conn.ssl.SSLConnectionSocketFactory] Starting handshake
11:56:51.380 [LIFECYCLE] [org.gradle.launcher.daemon.server.Daemon]
11:56:51.380 [DEBUG] [org.gradle.launcher.daemon.server.Daemon] DaemonExpirationPeriodicCheck running
11:56:51.384 [DEBUG] [org.gradle.cache.internal.DefaultFileLockManager] Waiting to acquire shared lock on daemon addresses registry.
11:56:51.385 [DEBUG] [org.gradle.cache.internal.DefaultFileLockManager] Lock acquired on daemon addresses registry.
11:56:51.385 [DEBUG] [org.gradle.cache.internal.DefaultFileLockManager] Releasing lock on daemon addresses registry.
11:56:51.387 [DEBUG] [org.gradle.cache.internal.DefaultFileLockManager] Waiting to acquire shared lock on daemon addresses registry.
11:56:51.387 [DEBUG] [org.gradle.cache.internal.DefaultFileLockManager] Lock acquired on daemon addresses registry.
11:56:51.388 [DEBUG] [org.gradle.cache.internal.DefaultFileLockManager] Releasing lock on daemon addresses registry.
11:57:01.381 [DEBUG] [org.gradle.launcher.daemon.server.Daemon] DaemonExpirationPeriodicCheck running
11:57:01.384 [DEBUG] [org.gradle.cache.internal.DefaultFileLockManager] Waiting to acquire shared lock on daemon addresses registry.
```
**Additional Information**: 

Because of limitations in my environment, I cannot access the Gradle plugin, so I am forced to use Jib by creating a custom task within Gradle using the Jib APIs. I am simply trying Jib out as a proof of concept by taking a base image, doing nothing with it, and publishing the "result" to a local insecure registry. When I do this, it hangs. Any advice on where I have misconfigured Jib or the registry is appreciated.

**Comment by chanseokoh** [comment]
Hi @neilchaudhuri,

As you can see in the log `GET https://localhost:5000/v2/`, Jib always tries HTTPS, even if `allowInsecureRegistries` is set. My guess is that the attempting `localhost:5000` with HTTPS just hangs. What happens when you do `curl https://localhost:5000/v2/`? (Note `https://`, not `http://`.) Does it hang too?

Also, does `docker push localhost:5000/...`? I am asking this because according to the Docker doc, it attempts HTTPS first before falling back to HTTP just like Jib.

And that also makes me wonder how you set up your local registry. `docker run -d -p 5000:5000 --restart=always --name registry registry:2` as in [here](https://docs.docker.com/registry/deploying/)? For me, it works fine locally. If you were using the same command,

**Comment by neilchaudhuri** [comment]
Thanks for your prompt response.

I noticed that in the logs too, but while I was confused why Jib would even try HTTPS even with the flag set, I had hoped that it would fall back to HTTP after failure or timeout. I guess not.

You are right that the curl hangs.

On the other hand, `docker push` works:

```
docker push localhost:5000/test/test-jib
The push refers to repository [localhost:5000/test/test-jib]
bf3bf7364c42: Pushed
0d6df433a894: Pushed
2847d17c4db8: Pushed
6e3181b1e193: Pushed
23e42653cb30: Pushed
cfb61419a39e: Pushed
be4626a574d8: Pushed
15d57950dad3: Pushed
2bf534399aca: Pushed
1c95c77433e8: Pushed
latest: digest: sha256:f1f499a48fca7e977034c32a2ae83e296fc56da21229ce1b392333b3196f4c82 size: 2419
```

I set up the registry precisely as described. As I am new-ish to Docker, your guidance is appreciated.

**Comment by TadCordle** [comment]
FYI, Jib's latest version is currently 1.5.1. I see you're on 0.10.1 which is pretty old, so maybe you should upgrade if possible (although I see you mentioned environment limitations).

**Comment by loosebazooka** [comment]
@TadCordle I think it's jib core that they're using?

**Comment by TadCordle** [comment]
Ah right, my bad.

**Comment by chanseokoh** [comment]
> I had hoped that it would fall back to HTTP after failure

Jib does fall back to HTTP after failure. The problem seems to be that it hangs forever instead of failing on your machine. On my laptop, it fails immediately, so the local setup works for me. I don't get why localhost:5000 doesn't fail when attempting HTTPS in your case. As if the connection attempt never goes through at all. localhost:5000 doesn't speak HTTPS, so as soon as curl attempts HTTPS, the handshake should fail no matter what. Might be some firewall rule on your machine, I don't know.

That said, you will have no problem if you use a secure remote registry such as Docker Hub.

If you can somehow adjust your network so that HTTPS on localhost:5000 goes though and fail, Jib should start working. But it is also very easy to make the local registry speak HTTPS (untrusted) instead of HTTP.

```
# Generate your own certificate.
# It will ask a few things, but it doesn't matter much what you answer in your case.
$ openssl req \
     -newkey rsa:4096 -nodes -sha256 -keyout certs/domain.key \
     -x509 -days 365 -out certs/domain.crt

# Run with TLS enabled (listening on port 5000)
$ docker run -d \
  --name registry \
  -v $HOME/docker/registry/certs:/certs \
  -e REGISTRY_HTTP_ADDR=0.0.0.0:443 \
  -e REGISTRY_HTTP_TLS_CERTIFICATE=/certs/domain.crt \
  -e REGISTRY_HTTP_TLS_KEY=/certs/domain.key \
  -p 5000:443 \
  registry:2
```
You still need to set `allowInsecureRegistries`.

**Comment by neilchaudhuri** [comment]
Unfortunately I don't have the permissions to install certificates, but it is good in a way to know something weird is up with my machine. I will try setting my `Containerizer.to` to something "real" and HTTPS. Thanks.

**Comment by chanseokoh** [comment]
You don't have to install certificates. I have updated the example above to fix a few things. For example, you can generate some arbitrary certs under your home directory `$HOME/my-certs`

```
$ mkdir $HOME/my-certs
$ openssl req ... \
    -keyout $HOME/my-certs/domain.key \
    -out $HOME/my-certs/domain.crt
```

and let the registry pick those up

```
$ docker run -d \
    ...
    -v $HOME/my-certs:/certs
```

Don't be fooled by the following lines. `/certs/domain.crt` refers to the files *inside* the container where the registry program runs.

```
  -e REGISTRY_HTTP_TLS_CERTIFICATE=/certs/domain.crt \
  -e REGISTRY_HTTP_TLS_KEY=/certs/domain.key \
```

**Comment by chanseokoh** [comment]
This may be a different issue, but just in case, I'm mentioning here that building images with Jib >= 1.5.0 can actually hang due to #1970.

**Comment by TadCordle** [comment]
@neilchaudhuri Like @chanseokoh said, it may be a different issue, but jib-core v0.11.0 has been released with the fix for #1970, so it might be worth trying.

**Comment by neilchaudhuri** [comment]
While there are other unrelated matters, things are definitely faster now with the Docker Daemon. Thanks for letting me know.
